stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala

```diff
@@ -1082,7 +1082,47 @@ trait FlowOps[+Out, +Mat] {
    * @param onComplete a function that transforms the ongoing state into an optional output element
    */
   def statefulMap[S, T](create: () => S)(f: (S, Out) => (S, T), onComplete: S => Option[T]): Repr[T] =
-    via(new StatefulMap[S, Out, T](create, f, onComplete))
+    via(new StatefulMap[S, Out, T](create, f, onComplete).withAttributes(DefaultAttributes.statefulMap))
+
+  /**
+   * Transform each stream element with the help of a resource.
+   *
+   * The resource creation function is invoked once when the stream is materialized and the returned resource is passed to
+   * the mapping function for mapping the first element. The mapping function returns a mapped element to emit
+   * downstream. The returned `T` MUST NOT be `null` as it is illegal as stream element - according to the Reactive Streams specification.
+   *
+   * The `close` function is called only once when the upstream or downstream finishes or fails. You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Early completion can be done with combination of the [[takeWhile]] operator.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * You can configure the default dispatcher for this Source by changing the `akka.stream.materializer.blocking-io-dispatcher` or
+   * set it for a given Source by using [[ActorAttributes]].
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam R the type of the resource
+   * @tparam T the type of the output elements
+   * @param create function that creates the resource
+   * @param f function that transforms the upstream element and the resource to output element
+   * @param close function that closes the resource, optionally outputting a last element
+   * @since 1.1.0
+   */
+  def mapWithResource[R, T](create: () => R)(f: (R, Out) => T, close: R => Option[T]): Repr[T] =
```

Comment:
For close function
`Using for scala3`
`scala.util.control.Exception.ultimately for scala2`
Do we need to consider such usage case for this function?

Reply:
Would you like elaborate?

Reply:
Taking scala3 as an example, when I close or use a stream, I will write this code
```
Using (Source. from (File...)) {
  // Todo something
}
```
I won't implement a close function. as also to Java8 `try (resource) {}`
I hope Pekko can do the same. Can we use `Using feature` and remove the params  `close:R=>Option[T]`
I haven't seen the implementation principle of `Using in scala3` yet, so I'm not sure how to define it, maybe further optimization in another PR

Reply:
I see , that would be nice addition.
1. directly support for Java's AutoCloseable.
2. directly support for Scala's Using.

But I think those two doesn't related to this PR, because the nice feature of this operator is "optional emit with the resource/final signal".

But and what your suggestion can come up with another PR.

Reply:
I think we need a github discussion on this, there are lots of ways we can do automatic resource management i.e. anything from scala.util.Using (which is also available in Scala 2.12) to more extensive methods available in cats-effect/ZIO.

This is something that really needs someone with extensive knowledge to look into it because its actually incredibly difficult to guarantee that resources are cleaned up in all cases.

Reply:
+1 about a discussion - the API should stay consistent with existing APIs. I think Creator API is the one to use and we can add tools to convert other functions to Creator function in cases where the compiler does not do it automatically.

Reply:
There is no consistency, and consitency for javadsl, we should move away from the `FI` , `JAPI`, `japi.function` as more as possible, and stay close and consistency with other java native reactive library and java stream api. By this way ,we can bring more users on board.

These APIs exist due to their historical necessity and rationality, but they may not necessarily adapt to future development. We need to look at these issues with a developmental perspective.

Reply:
There is a disagreement. Should we initiate an email vote for reference? 
I used to mainly write Spark apps. From the perspective of Spark API design, they all have related APIs in Java, Scala, Python and R. The API function names are the same and the semantics are basically the same, but the signatures are not completely consistent. User habits are different, or we should provide some function overloading.

Reply:
If we are talking about `Using`, since afaict this isn't used anywhere else in Pekko we should honestly discuss this elsewhere.

Regarding the use of `org.apache.pekko.japi.function.*` vs `java.lang.*` (i.e. `Supplier` vs `Creator`) this is actually entirely consistent right now if you **_take into account the various environments** i.e.

* For `javadsl`, `org.apache.pekko.japi.function.*` is used anytime a supplied lambda could legitimately throw an exception. This is because `java.lang.*` `FunctionalInterface`'s cannot throw exceptions (`@SneakyThrow` does not count, it comes from lombok and its a hacky workaround). Otherwise if if a function should never throw an exception (i.e. creating a collection) then you use functional interfaces in `java.lang.*`.
* For `scaladsl` you do not need `org.apache.pekko.japi.function.*` since you instead use the standard Scala functions (i.e. `X => X`) because standard functions in Scala can throw exceptions. One difference here is that `scaladsl` (unlike `javadsl`) does not use `java.lang.*` functional interfaces for lambda's that shouldn't throw an exception however it can be argued that this is just due to the fact that throwing exceptions in Scala are generally frowned upon due to it not being idiomatic in Scala. If someone wants to change this and suggest using `java.lang.*` intefaces for functions that shouldn't throw in `scaladsl` as well, they can also suggest that as a community discussion separate from this PR.

Reply:
That's funny, why java developers now suddenly need a `org.apache.pekko.japi.function.*` to ship things? And scala developers do not need it.

Reply:
> That's funny, why java developers now suddenly need a `org.apache.pekko.japi.function.*` to ship things? And scala developers do not need it.

Because Scala functions by default work like the `@FunctionalInterfaces` in `org.apache.pekko.japi.function.*`. In scala if you write a function like `def someMethod(lamba: () => ()) = ???` I can easily throw an exception in the supplied argument i.e. `someMethod(() => throw new Exception())`.

Its not that Scala developers don't need it, its that its not necessary because thats how Scala works by default.

Reply:
Could we consider using Pekko functions in javadsl for consistency and creating a new javadsl2 package which has standard Java functions instead? This is probably a big job but I don't think we can just change javadsl existing functions - and new functions should conform to the pre-existing style.

Reply:
Let's wait a little longer, I will change to `creator/callable` api this week, but I have to say, Java developers want `java.util.*`.

Reply:
Aggre with pjfanning. This PR should be more conservative. Annother PR will discuss the new Java APIs or method-overload to more confortable for java developer.

Reply:
> but I have to say, Java developers want `java.util.*`.

That may be the case (although I have no idea what the technical reason why is) but its outside the scope of this PR. Feel free to make a github discussion on it though!